services/mlmodel/tflite_cpu/tflite_cpu.go

```diff
@@ -0,0 +1,250 @@
+		if err != nil {
+			return nil, errors.Wrapf(err, "error getting parameters from config")
+		}
+		params, ok := tfParams.(*TFLiteConfig)
+		if !ok {
+			return nil, utils.NewUnexpectedTypeError(params, tfParams)
+		}
+		return params, nil
+	}, &TFLiteConfig{})
+}
+
+// TFLiteConfig contains the parameters specific to a tflite_cpu implementation
+// of the MLMS (machine learning model service).
+type TFLiteConfig struct {
+	// this should come from the attributes of the tflite_cpu instance of the MLMS
+	ModelPath  string  `json:"model_path"`
+	NumThreads int     `json:"num_threads"`
+	LabelPath  *string `json:"label_path"`
+}
+
+// Model is a struct that implements the TensorflowLite CPU implementation of the MLMS.
+// It includes the configured parameters, model struct, and associated metadata.
+type Model struct {
+	attrs    TFLiteConfig
+	model    *inf.TFLiteStruct
+	metadata *mlmodel.MLMetadata
+}
+
+// NewTFLiteCPUModel is a constructor that builds a tflite cpu implementation of the MLMS.
+func NewTFLiteCPUModel(ctx context.Context, params *TFLiteConfig) (mlmodel.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::addTFLiteModel")
+	defer span.End()
+	var model *inf.TFLiteStruct
+	var loader *inf.TFLiteModelLoader
+	var err error
+
+	addModel := func() (*inf.TFLiteStruct, error) {
+		if params == nil {
+			return nil, errors.New("could not find parameters")
+		}
+		if params.NumThreads <= 0 {
+			loader, err = inf.NewDefaultTFLiteModelLoader()
+		} else {
+			loader, err = inf.NewTFLiteModelLoader(params.NumThreads)
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get loader")
+		}
+
+		fullpath, err2 := fp.Abs(params.ModelPath)
+		if err2 != nil {
+			model, err = loader.Load(params.ModelPath)
+		} else {
+			model, err = loader.Load(fullpath)
+		}
+
+		if err != nil {
+			if strings.Contains(err.Error(), "failed to load") {
+				if err2 != nil {
+					return nil, errors.Wrapf(err, "file not found at %s", params.ModelPath)
+				}
+				return nil, errors.Wrapf(err, "file not found at %s", fullpath)
+			}
+			return nil, errors.Wrap(err, "loader could not load model")
+		}
+		return model, nil
+	}
+	model, err = addModel()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not add model from location %s", params.ModelPath)
+	}
+
+	return &Model{attrs: *params, model: model}, nil
+}
+
+// Infer takes the input map, finds the image (labeled 'image'), and uses the inference
+// package to return the result from the tflite cpu model as a map.
+func (m *Model) Infer(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error) {
+	_, span := trace.StartSpan(ctx, "service::mlmodel::tflite_cpu::Infer")
+	defer span.End()
+
+	outMap := make(map[string]interface{})
+	doInfer := func(input interface{}) (map[string]interface{}, error) {
+		outTensors, err := m.model.Infer(input)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't infer from model")
```

Comment:
If Infer fails, also print the name of the model that failed, m.Name (so store the name in the Model struct)